### ISSUE TITLE
ATTENTION, LEARN TO SOLVE ROUTING PROBLEMS!

### ISSUE THREAD

**Issue by Kazuki-115** [Kazuki-115/papers-savey#2]
https://arxiv.org/abs/1803.08475

**Comment by Kazuki-115** [comment]
# 提案と目的
- PointerNetworkに代わるAttention構造を用いたモデルの提案
- 貪欲ロールアウトに基づくシンプルなベースラインを用いたREINFORCE
- TSPに特化したアルゴリズム(Concordeなど)をoutperformするのではなく、さまざまな問題に対する機械学習手法の柔軟性を示すことが目的

# 先行研究
- NNの応用の最初はHopfield-networkを使って小さなTSPを解いたこと
- PointerNetworkの提案
	- attention機構を用いて入力のpermutationを出力する
- 正解データを除いてActor-CriticアルゴリズムでPointerNetworkを学習
- Encoder-Decoderの2層ではなく、Graph embeddingに基づいたシングルモデル
- Graph Neural Networkを用いて直接経路を出力
	- 上記の手法よりも精度は悪い
# 提案手法
- TSPに対するAttention Modelを提案
	- 他の問題に対してはinput・mask・decoder contextを定義し直す必要
- encoder は全ての入力nodeの埋め込みを作成
- decoderは入力nodeのsequenceを１つずつ出力
 - 一度構築された部分的なツアーは変更できない
	 - remaining problemは、最後のnode---unvisited node---最初のnodeをつなぐパスを見つけること
## Encoder
 - transformerに似ているが、positional encodingは使わない（入力の順番はどうでもいいから)
 - 2次元の入力から $d_h$次元の埋め込みを作る
	 - $h_i^0=W^xx_i+b^x$
- 埋め込みはN層のAttenitonによってupdate
### Attention Layer
- Multi-Head Attention(MHA) 
	- node間の相互作用・依存性の学習
- Fully Connected Feed-Forward(FF)
	- 各ノードの埋め込みを個別に深く処理
- 各sub layer はスキップ接続とbatch normalizationを追加
## Decoder
- 以下の2つに基づき時刻tごとにnode $\pi_t$ を出力
	- Enciderからの埋め込み
	- output $\pi_t'$ ($t'<t$)
- 各時刻tのdecoderの処理
	- 文脈ベクトルの作成
		- $h_{(c)}^{(N)}=[\bar{h}^{(N)},h_{\pi_{t-1}}^{(N)},h_{\pi_{2}}^{(N)}]$  
	- MHAで上記と各nodeの埋め込みベクトルを合算 --> $h_{c}^{(N+1)}$
	- 文脈ベクトルと各nodeの埋め込みベクトルからAttentionを計算
		- この時、過去に訪問した頂点のスコアは-infに
	- softmaxを通して確率に変換し、頂点を1つ選ぶ

##  REINFORCEのベースライン設計
- ベースラインの目的
	- 勾配推定のばらつきを小さくする
	- 安定してハイスピードな学習を実現
- 典型的なベースライン設計
	- 指数移動平均のベースライン
		- インスタンスごとに対応できない
	- Critic
		- 学習が難しい
- 提案するベースライン：Rollout Baseline
	- これまでのbest modeで定義されたポリシーから貪欲に決定したsolution cost
# 結果
- AM- greedy は求解速度がかなり早い
- 精度という観点だと最適解を出すことはできないが普通のヒューリスティックよりは良さそう
- 問題によってはGurobiよりも精度が高く本手法がかなり優位性を持つこともある
# 前提知識
- Multi Head Attention
- Attention計算
- skip-connection(He et al.,2016)
- batch normalization(Ioffe & Szegedy 2015)
	- layer noralizaton(Ba et al.,2016)よりも精度が良いらしい
- Reinforceのベースライン設計
